operationsgateway_api/src/functions/builtins/background.py

```diff
@@ -0,0 +1,54 @@
+import numpy as np
+
+from operationsgateway_api.src.functions.builtins.builtin import Builtin
+from operationsgateway_api.src.functions.waveform_variable import WaveformVariable
+
+
+class Background(Builtin):
+    input_types = {"waveform", "image"}
+    output_type = "scalar"
+    symbol = "background"
+    token = {
+        "symbol": symbol,
+        "name": "Background",
+        "details": (
+            "Calculate the background of a waveform or image. Errors if scalar "
+            "provided. "
+            "Implementation (waveform): First, applies smoothing by taking "
+            "weighted nearest and next-nearest neighbour contributions to y "
+            "values whose difference from their neighbours is more than 0.2 "
+            "times the total range in y. The first 25 and last 25 y values in "
+            "the signal are averaged to give an estimate of the background."
+            "Implementation (image): The average pixel value in the 10 by 10 region "
+            "in the top left of the image is returned."
+        ),
+    }
+
+    @staticmethod
+    def evaluate(argument: "WaveformVariable | np.ndarray") -> float:
+        """
+        Waveform:
+        First, applies smoothing by taking weighted nearest and next-nearest
+        neighbour contributions to  y values whose difference from their neighbours
+        is more than 0.2 times the total range in y.
+        The first 25 and last 25 y values in the signal are averaged to give an
+        estimate of the background.
+
+        Image:
+        The average pixel value in the 10 by 10 region in the top left of the image
+        is returned.
+        """
+        if isinstance(argument, WaveformVariable):
+            Builtin.smooth(argument.y)
+            if len(argument.y) < 50:
+                return np.mean(argument.y)
+            else:
+                return (np.mean(argument.y[:25]) + np.mean(argument.y[-25:])) / 2
+        elif isinstance(argument, np.ndarray):
+            return np.mean(argument[:10, :10])
+        else:
+            message = (
+                "'background' accepts [WaveformVariable, np.ndarray], "
```

Comment:
If things work as intended, then the user would see (or at least we'd provide the frontend with) the message generated by:
https://github.com/ral-facilities/operationsgateway-api/blob/035f38b5a033f2af2eba3ed32b7b722f7559eced/operationsgateway_api/src/functions/builtins/builtins.py#L41-L60
As we would type check functions explicitly before evaluation. There's a bit more effort in mapping the type to something human readable (obviously not for the types we accept, but would need to try and map the type actually provided by the user). But can definitely be done like the type check message here as well.
It's also worth mentioning that functions like FWHM won't throw explicit errors here (but would still error if you gave them the wrong type), it's only because `background` and `integrate` can apply to either WF or images that we're doing the type check.

Reply:
Actually this would only happen in places where we call `Background.evaluate` directly, usually it would be done by https://github.com/ral-facilities/operationsgateway-api/blob/035f38b5a033f2af2eba3ed32b7b722f7559eced/operationsgateway_api/src/functions/builtins/builtins.py#L62-L67
Which also does the above, human readable type check.